### PR TITLE
Use a VSCode dev container for local development

### PR DESCRIPTION
**Pull request by steersbob**
## Remote containers

To standardize the development environment for editing and building code, we can utilize a VSCode feature: remote development in a Docker container (https://code.visualstudio.com/docs/remote/containers).

The development container is defined in `.devcontainer/`. The Dockerfile installs the required toolchains for Particle, ESP-IDF, and Protobuf.
As this container is intended for development machines instead of CI builds, it is not uploaded to Docker Hub, but built locally.

When opening VSCode, you will be prompted to reload the editor in a container. If this does not happen, you may need to install the [Remote - Containers plugin](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers).
On first load, it will build the Docker image. This will take a while (5-30 minutes, depending on your machine).

## Configuration changes

The `.vscode/c_cpp_properties.json` file has been updated to provide best-effort intellisense results for all toolchains. This somewhat fudges the details, as it is hardcoded to use the host G++ compiler to avoid include errors for i386 (Particle) or xtensa-esp32 (ESP) system headers.
This is editor-only configuration, and does not reflect build settings.

The `docker/docker-compose.yml` compiler configuration has been removed, and build scripts that used it have been updated. All build scripts and tasks except the cross-compilation scripts (`build/build-sim-arm{32,64}.sh`) now use the locally available toolchain.

The ESP-IDF toolchain is not sourced by default (`/opt/esp/idf/export.sh`), as it updates PATH to override the system Python with its own virtualenv.
The Protobuf/nanopb compiler toolchain relies on the `protobuf` Python package that is not present here.
Intellisense does not rely on the export.sh script being sourced, and the `build/idf.sh` sources export.sh before running commands. In combination this means that it's not required to have the toolchain permanently activated while the dev container is active.

### CODE REVIEW

**Comment by steersbob** [comment]
@elcojacobs @CvRXX could you have a look whether this works out of the box, and whether any features are missing from the container?

If everything is present and understandable, I'll update the README to include setup instructions for new users.

**Comment by CvRXX** [comment]
I've tried it with the following remarks

### Container
I was missing the remote container plugin but after I installed it the container was detected.


### ESP
When running the ESP build the `signature_verification_key.bin` was missing. I know how to solve this but this might not be clear to a new developer. Maybe it could be good if a dummy key was generated if no key is provided. We could add this to the `CMakeLists.txt`

Autocomplete en development tools seem to work. 

There is not a task for flashing yet but I think there are some challenges cause the esp would need to be connected to the docker container. I'm not sure how trivial that is. 
### GCC
 - When running the task `build GCC` the following error occurred:
 ```bear: Internal error.
Traceback (most recent call last):
  File "/usr/bin/bear", line 289, in wrapper
    return function(*args, **kwargs)
  File "/usr/bin/bear", line 316, in intercept_build
    exit_code, current = capture(args, tools)
  File "/usr/bin/bear", line 340, in capture
    exit_code = run_build(args.build, env=environment)
  File "/usr/bin/bear", line 213, in run_build
    exit_code = subprocess.call(command, *args, **kwargs)
  File "/usr/lib/python3.8/subprocess.py", line 340, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/usr/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.8/subprocess.py", line 1704, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '--'
bear: Please run this command again and turn on verbose mode (add '-vvvv' as argument).
The terminal process "bash '../wrapped_make.sh', 'APP=brewblox', 'PLATFORM=p1'" failed to launch (exit code: 1).```

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -22,8 +24,11 @@
   ],
   // Use 'forwardPorts' to make a list of ports inside the container available locally.
   // "forwardPorts": [],
-  // Uncomment to use the Docker CLI from inside the container. See https://aka.ms/vscode-remote/samples/docker-from-docker.
-  // "mounts": [ "source=/var/run/docker.sock,target=/var/run/docker.sock,type=bind" ],
+  "mounts": [
+    "source=/var/run/docker.sock,target=/var/run/docker.sock,type=bind", // mount the Docker socket
+    "source=/etc/timezone,target=/etc/timezone,readonly,type=bind", // mount timezone info
```
Comment (CvRXX) [inline .devcontainer/devcontainer.json:29]: This is a problem. `/etc/timerzone` is afaik not standard (https://unix.stackexchange.com/questions/452559/what-is-etc-timezone-used-for). Is it possible to use `/etc/localtime`?

Comment (CvRXX) [reply .devcontainer/devcontainer.json:29]: If I remove the `/etc/timezone` mount the timezone still seems to be correct

Comment (steersbob) [reply .devcontainer/devcontainer.json]: Removed the /etc/timezone mount as discussed.

**Pull request merged**